### PR TITLE
efivars: retry open and read operations

### PR DESCRIPTION
**Pull request by keszybz**
On my laptop (Lenovo X1carbo 4th) I very occasionally see test-boot-timestamps
fail with this tb:

262/494 test-boot-timestamps                    FAIL    0.7348453998565674 s (killed by signal 6 SIGABRT)

08:12:48 SYSTEMD_LANGUAGE_FALLBACK_MAP='/home/zbyszek/src/systemd/src/locale/language-fallback-map' SYSTEMD_KBD_MODEL_MAP='/home/zbyszek/src/systemd/src/locale/kbd-model-map' PATH='/home/zbyszek/src/systemd/build:/home/zbyszek/.local/bin:/usr/lib64/qt-3.3/bin:/usr/share/Modules/bin:/usr/condabin:/usr/lib64/ccache:/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin:/home/zbyszek/bin:/var/lib/snapd/snap/bin' /home/zbyszek/src/systemd/build/test-boot-timestamps
--- stderr ---
Failed to read $container of PID 1, ignoring: Permission denied
Found container virtualization none.
Failed to get SystemdOptions EFI variable, ignoring: Interrupted system call
Failed to read ACPI FPDT: Permission denied
Failed to read LoaderTimeInitUSec: Interrupted system call
Failed to read EFI loader data: Interrupted system call
Assertion 'q >= 0' failed at src/test/test-boot-timestamps.c:84, function main(). Aborting.

Normally it takes ~0.02s, but here there's a slowdown to 0.73 and things fail with EINTR.
This happens only occasionally, and I haven't been able to capture a strace.

It would be to ignore that case in test-boot-timestamps or always translate
EINTR to -ENODATA. Nevertheless, I think it's better to retry, since this gives
as more resilient behaviour and avoids a transient failure.

### CODE REVIEW

**Comment by poettering** [comment]
isn't this a kernel bug? should we really tape over this? I don't think open() or read() on efivarfs should ever result on EINTR. @mjg59 any idea? (i think you wrote efivarfs originally, no?)

**Comment by poettering** [comment]
Oh, wtf is this:

https://github.com/torvalds/linux/blob/master/fs/efivarfs/file.c#L75

**Comment by poettering** [comment]
https://github.com/torvalds/linux/commit/bef3efbeb897b56867e271cdbc5f8adaacaeb9cd

What a terrible hack!

File: src/basic/efivars.c [hunk 1]
```diff
@@ -56,17 +61,25 @@ int efi_get_variable(
...
                         return -errno;
 
                 return 0;
         }
 
-        fd = open(p, O_RDONLY|O_NOCTTY|O_CLOEXEC);
-        if (fd < 0)
-                return -errno;
+        for (unsigned try = 0;; try++) {
+                fd = open(p, O_RDONLY|O_NOCTTY|O_CLOEXEC);
```
Comment (poettering) [inline src/basic/efivars.c]: doing the loop around open() is unnecessary apparently, see the kernel patch that is causing this, which only added this to read()

**Comment by mjg59** [comment]
Not my fault, I'm afraid :( Retrying read() seems like the best plan.

**Comment by keszybz** [comment]
Thanks for the investigation. I assumed some firmware issue and didn't think of checking the kernel.

Updated to only loop around read.

**Comment by poettering** [comment]
Please add comment to the sources explaining what is going on, with reference to the kernel patch making this necessary.

Maybe propagate an error such as EBUSY or ETIME or so when we give up, instead of ENODATA? After all we know the data is there, but just a bit too busy for us to read.

**Comment by keszybz** [comment]
Updated. I changed the return value to EBUSY. As long as the loop is enough to avoid seeing this in tests, the exact code shouldn't matter.

**Comment by poettering** [comment]
This started showing up because of 84c048799a78a121c0536bf947b02c99093d3bfc I am sure btw, which keeps checking efivarfs

**Comment by poettering** [comment]
also see #14828

**Pull request merged**